In [5]:
import matplotlib
matplotlib.use('TkAgg')
from pylab import *
import networkx as nx
import random as rd

def initialize():
    global g, time_steps
    g = nx.karate_club_graph()
    g.pos = nx.spring_layout(g)
    for i in g.nodes:
        g.nodes[i]['state'] = 1 if random() < .2 else 0 # originally, .5
    time_steps = 0

def observe():
    global g
    cla()
    nx.draw(g, vmin = 0, vmax = 1,
            node_color = [g.nodes[i]['state'] for i in g.nodes],
            pos = g.pos)

def update_original():
    global g, time_steps
    
    listener = rd.choice(list(g.nodes))
    speaker = rd.choice(list(g.neighbors(listener)))
    g.nodes[listener]['state'] = g.nodes[speaker]['state']
    
    time_steps +=1
    
    state = sum([g.nodes[i]['state'] for i in g.nodes])
    
    # checks if the network converged, i.e. all 0s or all 1s
    if state == 0 or state == len(list(g.nodes)):
        return time_steps    
    return False
    
def update_reversed():
    global g, time_steps
    
    speaker = rd.choice(list(g.nodes))
    listener = rd.choice(list(g.neighbors(speaker)))
    g.nodes[listener]['state'] = g.nodes[speaker]['state']
    
    time_steps +=1
    state = sum([g.nodes[i]['state'] for i in g.nodes])
    
    # checks if the network converged, i.e. all 0s or all 1s
    if state == 0 or state == len(list(g.nodes)):
        return time_steps    
    return False

def update_edges():
    global g, time_steps

    edge = rd.choice(list(g.edges))
    speaker,listener = rd.sample(edge, 2)
    g.nodes[listener]['state'] = g.nodes[speaker]['state']
    
    time_steps +=1
    state = sum([g.nodes[i]['state'] for i in g.nodes])
    
    # checks if the network converged, i.e. all 0s or all 1s
    if state == 0 or state == len(list(g.nodes)):
        return time_steps    
    return False
    
#     for i in g.nodes:
#         if g.nodes[i]['state'] == 0:
#             count_0 += 1
#             if count_0 == len(g.nodes):
#                 time += 1
#                 print('Node ', i, 'Time passed ', time) 
#                 break 
#                 self.started = False 
#         else:
#             count_1 += 1
#             if count_1 == len(g.nodes):
#                 time += 1
#                 print('Node ', i, 'Time passed ', time) 
#                 break
#                 self.stop_button
#             else:
#                 time += 1
    
interactive = False

def experimenter(num_sims):
    total_steps = 0
    for _ in range(num_sims):
        initialize()
        while not update_edges(): # plug in the update mode to test
            pass
        total_steps += time_steps
    
    return round(total_steps/num_sims)  

num_sims = 100

if interactive:
    # if not running tests, run the regular sim visualization
    import pycxsimulator
    pycxsimulator.GUI().start(func=[initialize, observe, update_original])
else:
    print('Average time to consensus: ', experimenter(num_sims))


Average time to consensus:  839


Original: Average time to consensus is 536 steps.
Reversed: Average time to consensus 2441 steps.
Edge-base update: Average time to consensus 839 steps.